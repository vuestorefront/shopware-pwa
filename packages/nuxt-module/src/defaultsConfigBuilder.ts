import { mergeWith, set, get } from "lodash";
import { getDefaultApiParams } from "@shopware-pwa/composables";
function _customizer(objValue: Object, srcValue: string[]) {
  if (Array.isArray(objValue)) {
    return [...new Set([...objValue, ...srcValue])];
  }
}

function _getDefaultConfig() {
  return JSON.parse(JSON.stringify(getDefaultApiParams()));
}

let finalConfig = _getDefaultConfig();
export default function defaultsConfigBuilder() {
  return {
    /**
     * Add value to existing config. Objects and arrays are merged recursively.
     * Primitives are overridden (example limit property).
     */
    add: (key: string, config: unknown) => {
      const property = get(finalConfig, key);
      if (!property) {
        set(finalConfig, key, config);
      } else if (Array.isArray(property)) {
        const newArray = Array.isArray(config) ? config : [config];
        const combinedArray = [...property, ...newArray];
        set(finalConfig, key, [...new Set(combinedArray)]);
      } else {
        mergeWith(property, config, _customizer);
      }
      return defaultsConfigBuilder();
    },
    /**
     * Replace config for specific key with given value.
     */
    replace(key: string, value: unknown) {
      set(finalConfig, key, value);
      return defaultsConfigBuilder();
    },
    /**
     * Remove config value for specific key.
     */
    remove(key: string, value?: unknown) {
      if (value) {
        const property = get(finalConfig, key);
        if (Array.isArray(property)) {
          const newValue = property.filter((item) => item !== value);
          set(finalConfig, key, newValue);
          return defaultsConfigBuilder();
        }
      }
      set(finalConfig, key, undefined);
      return defaultsConfigBuilder();
    },
    /**
     * Reset the config to the default. Do not use this method unless you know what you are doing.
     */
    _resetToDefault: () => {
      finalConfig = _getDefaultConfig();
    },
    /**
     * Get config for specific key. If not specified whole config will be returned.
     */
    get: (key?: string) => {
      return key ? get(finalConfig, key) : finalConfig;
    },
  };
}
