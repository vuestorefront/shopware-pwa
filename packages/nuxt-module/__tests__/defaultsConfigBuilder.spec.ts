import defaultsConfigBuilder from "../src/defaultsConfigBuilder";

describe("nuxt-module - defaultsConfigBuilder", () => {
  beforeEach(() => {
    defaultsConfigBuilder()._resetToDefault();
  });

  describe("get", () => {
    it("should return default config keys", () => {
      const result = defaultsConfigBuilder().get();
      expect(Object.keys(result)).toMatchSnapshot();
    });

    it("should return cofig for specific key", () => {
      const result = defaultsConfigBuilder().get("useCms.limit");
      expect(result).toEqual(10);
    });
  });

  describe("add", () => {
    it("should replace value existing key", () => {
      const result = defaultsConfigBuilder().add("useCms", { limit: 3 }).get();
      expect(result.useCms.limit).toEqual(3);
    });

    it("should add new key to array without duplication", () => {
      const result = defaultsConfigBuilder()
        .add("useCms", {
          includes: { cms_page_slot: ["something", "type"] },
        })
        .get();
      expect(result.useCms?.includes?.cms_page_slot).toEqual([
        "id",
        "type",
        "slot",
        "blockId",
        "config",
        "data",
        "backgroundMediaMode",
        "backgroundMedia",
        "something",
      ]);
    });

    it("should add association", () => {
      const result = defaultsConfigBuilder()
        .add("useCms", {
          associations: { manufacturer: { associations: { something: {} } } },
        })
        .get();
      expect(result.useCms?.associations?.manufacturer).toEqual({
        associations: { media: {}, something: {} },
      });
    });

    it("should add new key to config", () => {
      const result = defaultsConfigBuilder()
        .add("myNewKey", {
          someConfig: 123,
        })
        .get();
      expect(result.myNewKey).toEqual({
        someConfig: 123,
      });
    });

    it("should not remove unit price with direct key", () => {
      const testKey = "useCms.includes.calculated_price";
      const previousValue = defaultsConfigBuilder().get(testKey);
      const result = defaultsConfigBuilder().add(testKey, ["referencePrice"]);
      expect(result.get(testKey)).toEqual([...previousValue, "referencePrice"]);
    });

    it("should not remove unit price without direct key", () => {
      const testKey = "useCms.includes.calculated_price";
      const previousValue = defaultsConfigBuilder().get(testKey);
      const result = defaultsConfigBuilder().add("useCms", {
        includes: { calculated_price: ["referencePrice"] },
      });
      expect(result.get(testKey)).toEqual([...previousValue, "referencePrice"]);
    });
  });

  describe("replace", () => {
    it("should replace value", () => {
      const result = defaultsConfigBuilder()
        .replace("useCms", { limit: 3 })
        .get();
      expect(result.useCms).toEqual({ limit: 3 });
    });

    it("should replace value by key", () => {
      const result = defaultsConfigBuilder().replace("useCms.limit", 5).get();
      expect(result.useCms?.limit).toEqual(5);
    });

    it("should replace value in array", () => {
      const result = defaultsConfigBuilder()
        .replace("useCms.includes.cms_page_slot", ["id", "type"])
        .get();
      expect(result.useCms?.includes?.cms_page_slot).toEqual(["id", "type"]);
    });
  });

  describe("remove", () => {
    it("should remove value", () => {
      const result = defaultsConfigBuilder().remove("useCms.limit").get();
      expect(result.useCms?.limit).toBeUndefined();
    });

    it("should remove array key", () => {
      const result = defaultsConfigBuilder()
        .remove("useCms.includes.cms_page_slot")
        .get();
      expect(result.useCms?.includes?.cms_page_slot).toBeUndefined();
    });

    it("should remove value from array by key", () => {
      const result = defaultsConfigBuilder()
        .remove("useCms.includes.cms_page_slot", "blockId")
        .get();
      expect(result.useCms?.includes?.cms_page_slot).not.toBeUndefined();
      expect(result.useCms?.includes?.cms_page_slot).not.toContain("blockId");
    });

    it("should ignore additional property when removing key property is not an array", () => {
      const result = defaultsConfigBuilder()
        .remove("useCms.limit", "notFromArray")
        .get();
      expect(result.useCms?.limit).toBeUndefined();
    });
  });

  it("should test the case from useDefaults documentation", () => {
    const result = defaultsConfigBuilder()
      .replace("useCms.limit", 8) // change default listing limit to 8
      .add("useCms.includes.product", "new_option") // add product new_option to returned fields
      .get();

    expect(result.useCms?.limit).toEqual(8);
    expect(result.useCms?.includes?.product).toContain("new_option");
  });
});
