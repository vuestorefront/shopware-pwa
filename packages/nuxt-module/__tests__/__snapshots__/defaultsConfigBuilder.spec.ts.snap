// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nuxt-module - defaultsConfigBuilder get should return default config keys 1`] = `
Array [
  "useCms",
  "useProductListing",
  "useProductQuickSearch",
  "useListing",
  "useProduct",
  "useNavigation",
  "useOrderDetails",
  "useCustomerOrders",
  "useCustomerAddresses",
  "useUser",
  "useCart",
]
`;
